### PR TITLE
Loose pin for openmpi and mpich

### PR DESCRIPTION
**Pull request by leofang**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by leofang** [comment]
cc: @isuruf @beckermr

File: recipe/gen_patch_json.py [hunk 1]
```diff
@@ -654,6 +654,12 @@ def _gen_new_index(repodata, subdir):
         if any(dep.startswith("zstd >=1.4") for dep in deps):
             _pin_looser(fn, record, "zstd", max_pin="x.x")
 
+        if any(dep.startswith("openmpi >=4.1") for dep in deps):
```
Comment (isuruf) [inline recipe/gen_patch_json.py:657]:
```suggestion
        if any(dep.startswith("openmpi >=4.0") for dep in deps):
```

Comment (leofang) [reply recipe/gen_patch_json.py]: 👍

**Review by beckermr — CHANGES_REQUESTED** [review]
Don't we want to adjust anything matching 4.x for openmpi and have no upper bound  for mpich.

**Comment by leofang** [comment]
Why no upper bound for mpich? Is its ABI always stable?!

**Comment by beckermr** [comment]
@isuruf told me it's abi is stable across major versions.

**Comment by beckermr** [comment]
Wait a second. Abi lab says it changed soname on 3.4. Does that not matter?

**Comment by isuruf** [comment]
> @isuruf told me it's abi is stable across major versions.

Sorry, I was wrong, but mpich hasn't bumped the major version in years.

**Comment by isuruf** [comment]
This looks good to me, but we need to fix the `run_exports` of `mpich` first.

**Comment by leofang** [comment]
Hmm 🤔 But it also says 100% compatible... I don't get it either

**Comment by leofang** [comment]
Oh yeah nvm. It's not even beta 😂

File: recipe/gen_patch_json.py [hunk 2]
```diff
@@ -654,6 +654,12 @@ def _gen_new_index(repodata, subdir):
         if any(dep.startswith("zstd >=1.4") for dep in deps):
             _pin_looser(fn, record, "zstd", max_pin="x.x")
 
+        if any(dep.startswith("openmpi >=4.0") for dep in deps):
```
Comment (leofang) [inline recipe/gen_patch_json.py:657]: @isuruf Wouldn't it be better if we just say 'openmpi>=4' here?

Comment (isuruf) [reply recipe/gen_patch_json.py:657]: Nope. We don't want to patch newer packages. Repodata patches should apply to  older ones and not new ones.

Comment (leofang) [reply recipe/gen_patch_json.py:657]: But what happens to openmpi 4.1.0?

**Comment by leofang** [comment]
> This looks good to me, but we need to fix the `run_exports` of `mpich` first.

https://github.com/conda-forge/mpich-feedstock/pull/54 is in.

**Comment by leofang** [comment]
@conda-forge/core this should be ready?

**Comment by leofang** [comment]
Thanks, @isuruf. How long does it take to be in effect?

**Comment by isuruf** [comment]
1 CDN sync (~ 1 hour)

**Comment by leofang** [comment]
OK thanks. I will check it later to see if https://github.com/conda-forge/mpi4py-feedstock/issues/36 is fixed.

**Comment by leofang** [comment]
Looks like we're good!

**Comment by beckermr** [comment]
I am seeing pins from 3.4 to 3.5 in mpi4py. I think we need to patch again.

**Comment by beckermr** [comment]
also this binary is not patched: osx-64/hdf5-1.10.5-mpi_mpich_hd288fd1_1014.tar.bz2

**Pull request merged**